lib/ood_core/job/adapters/ccq.rb

```diff
@@ -0,0 +1,300 @@
+          s.success? ? o : interpret_and_raise(e, cmd)
+        end
+
+        # helper function to interpret an error the command had given and
+        # raise a different error.
+        def interpret_and_raise(error, command)
+          # a special case with CCQ that prompts the user for username & password
+          # so let's be helpful and tell the user what to do.
+          if error.end_with?("EOFError: EOF when reading a line\n")
+            raise(
+              PromptError,
+              "The #{command} command was prompted. You need to generate the certificate " + 
+              "manually in a shell by running 'ccqstat'\nand entering your username/password"
+            )
+          else
+            raise(JobAdapterError, e.message)
+          end
+        end
+
+        # Convert seconds to duration
+        def seconds_to_duration(seconds)
+          format("%02d:%02d:%02d", seconds / 3600, seconds / 60 % 60, seconds % 60)
+        end
+
+        # helper to make a script file. We can't pipe it into ccq so we have to
+        # write a file.
+        def make_script_file(content)
+          file = Tempfile.new(tmp_file_name)
+          file.write(content.to_s)
+          file.flush
+          file
+        end
+
+        def tmp_file_name
+          'ccq_ood_script_'
+        end
+
+        def parse_job_id_from_ccqsub(output)
+          match_data = /job id is: (?<job_id>\d+) you/.match(output)
+          throw JobAdapterError.new "Could not extract job id out of ccqsub output '#{output}'" if match_data.nil? 
+          match_data[:job_id]
+        end
+
+        # parse an Ood::Job::Info object from extended ccqstat output
+        def info_from_ccqstat_extended(data)
+          raw = extended_data_to_hash(data)
+          data_hash = { native: raw }
+          data_hash[:status] = get_state(raw['status'])
+          data_hash[:id] = raw['name']
+          data_hash[:job_name] = raw['jobName']
+          data_hash[:job_owner] = raw['userName']
+          data_hash[:submit_host] = raw['submitHostInstanceId']
+          data_hash[:dispatch_time] = raw['startTime'].to_i
+          data_hash[:submission_time] = raw['dateSubmitted'].to_i
+          data_hash[:queue_name] = raw['criteriaPriority']
+
+          Info.new(data_hash)
+        end
+
+        # extended data is just lines of 'key: value' value, so parse
+        # it and stick it all in a hash.
+        def extended_data_to_hash(data)
+          header = true
+          data_hash = {}
+
+          data.to_s.each_line do |line|
+            unless header
+              words = line.split(':')
+              data_hash[words[0]] = words[1].gsub(/\s+/, "") if words.size == 2
+            end
+
+            header = false if header
+          end
+
+          data_hash
+        end
+
+        def info_from_ccqstat(data)
+          infos = []
+
+          data.to_s.each_line do |line|
+            words = line.split(/\s/).reject(&:empty?)
```

Comment:
Could the job name ever contain spaces? Can we specify the delimiter with ccqstat like we do with Slurm?

Reply:
Good catch. They absolutely can contain spaces (I just tested this).  I think the formatting is a bit off anyhow (my username `johrstrom` turns into `johrs...`) so we may be able to work with them on that for correctness.

Reply:
This is the only thing I can't solve immediately and quickly. It is well formatted from the header (see [this fixture](https://github.com/OSC/ood_core/blob/33088fdbe82844103ea4baac5a0b212d90285a95/spec/fixtures/output/ccq/ccqstat)) so we may be able to store the string index of the column. I'll reach out to the CCQ team about it.

Reply:
We can merge this and open an issue to address this issue.